### PR TITLE
Update README.md

### PR DESCRIPTION
**Pull request by awf**
Some proposed changes to README.  Food for discussion, and partially an aspirational direction of travel for ergonomics.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,19 +1,122 @@
...
+The benchmarks at TODO show that this library has performance highly competitive with (i.e. faster than) those and PyTorch on CPU.  
+It is hoped that the separation of AD logic from matrix and tensor manipulation (deferred to [hmatrix] and [orthotope]) will enable similar speedups on numerical accelerators.
+
+## Computing the derivative of a simple function
+
+Here is an example of computing the gradient of a function that goes from `R^3` to `R`:
+
+```Haskell
+-- A function that goes from `R^3` to `R`. 
+-- [Arguments are tupled, could just as easily be curried. -awf]
```
Comment (awf) [inline README.md:19]: These [-awf] comments are just for the pull request, so I didn't forget to make them.  They would be removed before merging.

**Comment by Mikolaj** [comment]
Thank you so much. Please take your time.

File: README.md [hunk 2]
```diff
@@ -1,19 +1,122 @@
...
+In order to compute the gradient, we transform the function in the classic manner for forward-mode AD, to take `DualNumber` variables, 
+and the `DualMonad` which implements the system in the paper
+```Haskell
+fooM :: DualMonad d r m 
+     => DualNumberVariables d r -> m (DualNumber d r) 
+fooM variables = do { 
+     let x : y : z : _ = vars variables 
+     w <- x * sin y 
+     returnLet $ (atan2 z w + z * w)  -- [note extra w, we *want* to require returnLet up front --awf] 
+   }
```
Comment (awf) [inline README.md:35]: Probably incorrect on every line - I am guessing, but I am fairly confident this should contain a `do`, a `<-` and a `returnLet`
I assume I should have written `sinM` and `atan2M` or similar?

File: README.md [hunk 3]
```diff
@@ -1,19 +1,122 @@
...
+```
+We hope to automate this transfomation soon, automatically generating `fooM` from `foo`, but for now it's a mechanical, if somewhat tedious, manual task.
+
+And then the gradient of `foo` is
+```Haskell
+grad_foo :: HasDelta r => Domain0 r -> (Domain0 r, r) 
+grad_foo ds = 
+   let ((result, _, _, _), value) = 
+         dReverse 1 fooM (ds, V.empty, V.empty, V.empty) 
+   in (result, value)
```
Comment (awf) [inline README.md:45]: Similarly - whatever is correct here.

File: README.md [hunk 4]
```diff
@@ -1,19 +1,122 @@
...
+grad_foo ds = 
+   let ((result, _, _, _), value) = 
+         dReverse 1 fooM (ds, V.empty, V.empty, V.empty) 
+   in (result, value) 
+```
+
+As can be verified by computing the gradient at `(1.1,2.2,3.3)`:
+```Haskell
+>>> grad_foo (mkDual [1.1, 2.2, 3.3])
+[7.75, 1.23, 5.52]
```
Comment (awf) [inline README.md:51]: This should remain a 2-liner, with as little line noise as possible. 
E.g. `grad_foo (DualVector.fromList [1.1 :: Float, 2.2, 3.3]`)

File: README.md [hunk 5]
```diff
@@ -1,19 +1,122 @@
...
-
-Here is an example of computing the gradient of a function that goes from `R^3` to `R^2`:
-
-https://github.com/Mikolaj/horde-ad/blob/40951dab17ae90c5ce528a59216a39a3727eccb5/test/common/TestSingleGradient.hs#L442-L516
-
-Elsewhere in the same file is a computation of the forward derivative of the function and a QuickCheck test relating it to the gradient. It uses the same definition of the objective function and the same glue code for grouping parameters, etc. The ratio of signal to noise (maths to glue code) is much higher in more complex functions, e.g., neural networks. Several sample neural networks for MNIST digit classification are gathered in
+# HORDE-AD: Higher Order Reverse Derivatives Efficiently 
+
+This is an Automatic Differentiation library based on the paper _"Provably correct, asymptotically efficient, higher-order reverse-mode automatic differentiation"_ by Krawiec, Krishnaswami, Peyton Jones, Ellis, Fitzgibbon, and Eisenberg.
```
Comment (awf) [inline README.md:3]: I think we should add something like:

> Primarily implemented by @Mikolaj directly from the paper, and then subsequently in collaboration with the authors

Comment (Mikolaj) [reply README.md:3]: Or we can skip myself (I'm credited in CREDITS.md). Regardless, mentioning the paper authors would be great, as it immediately elevates the library from the category of fan fiction to the category of a canonical piece.

BTW, it doesn't really matter, but it was implemented directly from Simon's talk (with snippets of Haskell code) and only a few details Simon omitted were later taken from the paper (which explains some of the cases where terminology not always follows the paper). However, by transitivity, it was from the paper.

File: README.md [hunk 6]
```diff
@@ -1,19 +1,122 @@
...
+```Haskell
+convMnistTwoS
+  :: forall kh kw num_hidden c_out h w c_in batch_size d r m.
+     ( KnownNat kh, KnownNat kw, KnownNat num_hidden, KnownNat c_out, KnownNat h, KnownNat w, KnownNat batch_size
+     , c_in ~ 1 -- greyscale
+     , 1 <= kh  -- kh greater than 1
+     , 1 <= kw  -- kw greater than 1
+     , DualMonad d r m )
+  => OS.Array '[batch_size, c_in, h, w] r
+  -> DualNumber d (OS.Array '[ c_out, c_in , kh + 1, kw + 1 ] r)
```
Comment (Mikolaj) [inline README.md:95]: To fit such things into a single line, perhaps the `DualNumber` type constructor could be renamed to `D` (the same as the value constructor of that type). That's what @tomjaguarpaw did in his variant of the implementation. `DualNumber` is less cryptic at the first glance, but more noisy in the long run, as we can see here.

File: README.md [hunk 7]
```diff
@@ -1,19 +1,122 @@
...
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ c_out, c_out, kh + 1, kw + 1 ] r)
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ num_hidden, c_out,
+                                 GHC.TypeLits.* ((h + kh) `Div` 2 + kh) `Div` 2
+                                 GHC.TypeLits.* ((w + kw) `Div` 2 + kw) `Div` 2 -- [buglet? kh->kw? -awf]
```
Comment (Mikolaj) [inline README.md:101]: Yes, this is a bug and it was not caught by GHC due to the not implemented `transpose2` TODO above (now it's caught). This beautifully undermines the whole story of correctness of shapes guaranteed by dependent types.

The flop was possible, because Haskell is not a dependently typed language, as Agda or Idris, but a wide-spectrum language, where you can mix dependent and normal typing. Unfortunately, it's not easy to detect if such mixing is taking place or not. The new `transpose2S` function is fully dependently typed, while the `from2S` used before is, in fact, a coercion of matrices to rank 2 tensors, so its very purpose is to override dependent typing.

In the code you cited, after the TODO is fixed, the only remaining lax function is `reshapeS` that, at most, could guarantee on the type level that the number of elements does not change after the tensor reshaping. However what the resulting ranks are is unknown and so the function is a coercion and can hide bugs. We could instead use function `flattenS1S` that would require that the resulting tensor is of rank 1 and so no flexibility would be left in the output of the function. That's probably worth doing just for the sake of this example. However, at some point we need to decide how many differentiable tensor manipulation operations we want to maintain and if it's not better, e.g.,  to constrain the use of `reshapeS` with `:: DualNumber d (OS.Array '[... * ... * ...])` instead of adding `flattenS1S` that does the very same thing but in one word.

File: README.md [hunk 8]
```diff
@@ -1,19 +1,122 @@
...
+Several sample neural networks for MNIST digit classification are gathered in
 
 https://github.com/Mikolaj/horde-ad/tree/master/src/HordeAd/Tool
 
 Among them are recurrent, convolutional and fully connected networks based on fully typed tensors (sizes of all dimensions are tracked in the types) as well as weakly typed fully connected networks built with, respectively, matrices, vectors and raw scalars (most flexible but slowest).
 
 
+# Derivative checking
+
+Elsewhere in the same file is a computation of the forward derivative of the function and a QuickCheck test relating it to the gradient. It uses the same definition of the objective function and the same glue code for grouping parameters, etc. The ratio of signal to noise (maths to glue code) is much higher in more complex functions, e.g., neural networks.
```
Comment (Mikolaj) [inline README.md:118]: This could be refined once we settle on the text above it, because by this point you've already shown a fragment of neural network definition, omitting the glue/noise. However, it's still very valuable IMHO that the QuickCheck tests exist for many small and big examples, including both examples in the README and many other mentioned in the README. And they are not smoke tests, but randomized checks of the fundamental formula you suggested

> dt <.> derivativeFromFun f ds = gradientFromFun dt f <.> ds

which the code was not designed to compute nor fine-tuned to pass. The tests pass as a happy coincidence, because the code is faithful to maths and they can be expressed as cleanly as above, because the API mimics maths.

**Review by Mikolaj — COMMENTED** [review]
Thank you, that's a great improvement. Over to you. :)

File: README.md [hunk 9]
```diff
@@ -1,19 +1,122 @@
...
+foo :: RealFloat a => (a,a,a) -> (a,a)
+foo (x,y,z) = 
+  let w = x * sin y 
+  in (atan2 z w, z * w)
+```
+TODO: show how the 2x3 Jacobian emerges from here
+
+# Forall shapes and sizes
+
+An additional feature of this library is a type system for tensor shape arithmetic.  Look at this beauty.
```
[the remark this answers is not in the document]
Comment (tomjaguarpaw) [reply README.md:67]: Isn't it true regardless of whether `orthotope` provides the heavy lifting for us?

File: README.md [hunk 10]
```diff
@@ -1,19 +1,122 @@
...
+     let x : y : z : _ = vars variables 
+     w <- x * sin y 
+     returnLet $ (atan2 z w + z * w)  -- [note extra w, we *want* to require returnLet up front --awf] 
+   }
+```
+We hope to automate this transfomation soon, automatically generating `fooM` from `foo`, but for now it's a mechanical, if somewhat tedious, manual task.
+
+And then the gradient of `foo` is
+```Haskell
+grad_foo :: HasDelta r => Domain0 r -> (Domain0 r, r)
```
[the remark this answers is not in the document]
Comment (tomjaguarpaw) [reply README.md:41]: Yeah, we need to abolish `DualNumberVariables` and `Domain_` in favour of something like `AdaptArg` that I have in my version. That way we can preserve the tuple structure of the inputs and outputs. Still, my `AdaptArg` itself needs to be made more palatable, but it's the right direction to go.

File: README.md (hunk 5 above)
Comment (Mikolaj) [reply README.md:3]: Right. It's finally published now.

File: README.md [hunk 11]
```diff
@@ -1,19 +1,122 @@
...
+# HORDE-AD: Higher Order Reverse Derivatives Efficiently 
+
+This is an Automatic Differentiation library based on the paper _"Provably correct, asymptotically efficient, higher-order reverse-mode automatic differentiation"_ by Krawiec, Krishnaswami, Peyton Jones, Ellis, Fitzgibbon, and Eisenberg.
+
+This is an early prototype, both in terms of the engine performance and the API and toolbox for the library user. The user should be ready to add missing primitives, as well as obvious tools that should be predefined but aren't. One can already build with the library all basic neural network architectures, such as fully connected, recurrent, convolutional and residual networks. The library, due to its loose coupling of differentiation and data containers, can naturally handle exotic variants of such networks that wouldn't express well in a language of matrices or tensors. 
+
+Applications outside machine learning are very much in scope, given that the notion of a neural network is not hardwired into the formalism, but ad hoc redefined from basic blocks of general automatic differentiation whenever it's needed.
```
[the remark this answers is not in the document]
Comment (Mikolaj) [reply README.md:7]: Basic blocks is, on one hand, various differentiable functions that already come with the library: arithmetic, tensor ops, loss and activation functions, various kinds of data or weight normalization (#42), convolution-related ops, etc.. Then there are mechanisms for defining new differentiable functions (not just by combining the ones bundled with the library). Then there are operations that compute gradient, derivative, etc. Then various optimizers (currently only variants of gradient descent). The last two depend on the iffy mechanism (`DualNumberVariables`) for bundling parameters of various ranks and sizes into one value that can be passed along, taken derivative or gradient with respect to, mutated in gradient descent iterations, etc.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mikolaj) [reply README.md:41]: Yes, that would be great (#48). I guess at rank 0 it would kill performance, but if so, we can keep both systems to be able to squeeze out the performance for that use case (a rare case, probably, but many toy benchmarks live there, including many of ours).

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mikolaj) [reply README.md:7]: BTW, would "basic building blocks" be more understandable?

File: README.md [hunk 12]
```diff
@@ -1,19 +1,122 @@
...
+-- A function that goes from `R^3` to `R`. 
+-- [Arguments are tupled, could just as easily be curried. -awf]
+foo :: RealFloat a => (a,a,a) -> a
+foo (x,y,z) = 
+  let w = x * sin y 
+  in atan2 z w + z * w  -- [note extra w, we *want* to require returnLet up front --awf] 
+```
+
+In order to compute the gradient, we transform the function in the classic manner for forward-mode AD, to take `DualNumber` variables, 
+and the `DualMonad` which implements the system in the paper
```
[the remark this answers is not in the document]
Comment (Mikolaj) [reply README.md:27]: > But I'm still concerned about the types! Do we really lose the compile-time check that calls to foo are expected to take three variables?

Yes, we do. When you look at big shaped examples in https://github.com/Mikolaj/horde-ad/tree/master/src/HordeAd/Tool, the creative part of the code is nice and typed and composes in a typed way and at the end comes the glue code that explicitly defines tensor shapes that are already implicitly given in the tuple of your function arguments, squashes types, adds proxies and makes it all a mess. This could be handled by the plugin that is supposed to add the monad things. Or we could rework the types of the "basic building blocks" of the library (optimizers, generators of random weights, etc.) so that they don't expect a single tuple of vectors, but something more clever, e.g., related to @tomjaguarpaw's biapplicative `AdaptArg`.

**Comment by Mikolaj** [comment]
This is now mostly done, with resulted discussed in #75 (please discuss some more). Thank you! Closing.
